robosherlock/src/io/include/rs/io/visualizer.h

```diff
@@ -70,41 +76,100 @@ class Visualizer
+  // api RS ProcessManager is calling this from the outside Note: When running in MultiAAE mode, this will only affect
+  // the first AAE (or in more detail: rs::Visualizer::visualizableGroupManagers_.begin().second
   void setActiveAnnotators(std::vector<std::string> annotators);
   std::string nextAnnotator();
   std::string prevAnnotator();
   std::string selectAnnotator(std::string annotator);
 
+  /** Add a VisualizableGroupManager for a given identifier.
+   * Only required in MultiAAE mode.
+   *
+   * Please use this method directly after creating a new RSAAE
+   * or instantiating your Visualizables in any other way.
+   * This create an VisualizableGroupManager that takes care of the drawing state
+   * of the different AAEs or other visualizing entities that might be run.
+   *
+   * @param identifier A unique identifier. This is usually the name of your AAE when you only execute AAEs.   */
+  void addVisualizableGroupManager(std::string identifier);
+
 private:
-  static void callbackMouse(const int event, const int x, const int y, const int flags, void *object);
-  void callbackMouseHandler(const int event, const int x, const int y);
-  void callbackKeyHandler(const char key, const DrawingAnnotator::Source source);
+  static void callbackMouse(const int event, const int x, const int y, const int flags, void* object);
+
+  /**
+   * This method will be called when the Visualizer detected a Keypress.
+   *  @param activeVGM The Associated VisualizableGroupManager in which window the detected key has been pressed
+   */
+  void callbackKeyHandler(const char key, const Visualizable::VisualizableDataType source,
+                          std::shared_ptr<VisualizableGroupManager> activeVGM);
 
-  void checkAnnotator();
   void shutdown();
 
   void imageViewer();
   void cloudViewer();
 
-  void keyboardEventImageViewer(const cv::Mat &disp);
-  void keyboardEventCloudViewer(const pcl::visualization::KeyboardEvent &event, void *);
-
-  void saveImage(const cv::Mat &disp);
-  void saveCloud(const pcl::PointCloud<pcl::PointXYZRGBA>::Ptr cloud, pcl::visualization::PCLVisualizer::Ptr &visualizer);
-
-  bool visControlCallback(robosherlock_msgs::RSVisControl::Request &req,
-      robosherlock_msgs::RSVisControl::Response &res);
-
+  void keyboardEventImageViewer(const cv::Mat& disp);
+  void keyboardEventCloudViewer(const pcl::visualization::KeyboardEvent& event, void*);
+
+  void saveImage(const cv::Mat& disp, std::shared_ptr<VisualizableGroupManager> vgm);
+  void saveCloud(const pcl::PointCloud<pcl::PointXYZRGBA>::Ptr cloud,
+                 pcl::visualization::PCLVisualizer::Ptr& visualizer);
+
+  inline const std::string imageWindowName(VisualizableGroupManager& vgm)
+  {
+    return vgm.getIdentifier() + "/Image Viewer";
+  }
+  inline const std::string cloudWindowName(VisualizableGroupManager& vgm)
+  {
+    return vgm.getIdentifier() + "/Cloud Viewer";
+  }
+
+  // https://stackoverflow.com/a/478960
```

Comment:
could you also add the gist of this post here pls. It has happened to me that the post got changed/removed etc.

Reply:
Done :)